mod/likes/actions/likes/add.php

```diff
@@ -1,99 +1,91 @@
-$annotation_id = create_annotation($entity->guid,
-								'likes',
-								"likes",
-								"",
-								$user->guid,
-								$entity->access_id);
+$annotation_id = create_annotation($entity->guid, 'likes', 'likes', '', $user->guid, $entity->access_id);
 
 // tell user annotation didn't work if that is the case
 if (!$annotation_id) {
-	register_error(elgg_echo("likes:failure"));
-	forward(REFERER);
+	return elgg_error_response(elgg_echo('likes:failure'));
 }
 
+if ($entity->owner_guid === $user->guid) {
+	return elgg_ok_response('', elgg_echo('likes:likes'));
+}
+
+// @todo move this notification to a notification event handler
 // notify if poster wasn't owner
-if ($entity->owner_guid != $user->guid) {
-	$owner = $entity->getOwnerEntity();
-
-	$annotation = elgg_get_annotation_from_id($annotation_id);
-
-	$title_str = $entity->getDisplayName();
-	if (!$title_str) {
-		$title_str = elgg_get_excerpt($entity->description, 80);
-	}
-
-	$site = elgg_get_site_entity();
-
-	// summary for site_notifications
-	$summary = elgg_echo('likes:notifications:subject', [
-			$user->name,
-			$title_str
-		],
-		$owner->language
-	);
-	
-	// prevent long subjects in mail
-	$title_str = elgg_get_excerpt($title_str, 80);
-	$subject = elgg_echo('likes:notifications:subject', [
-			$user->name,
-			$title_str
-		],
-		$owner->language
-	);
-
-	$body = elgg_echo('likes:notifications:body', [
-			$owner->name,
-			$user->name,
-			$title_str,
-			$site->name,
-			$entity->getURL(),
-			$user->getURL()
-		],
-		$owner->language
-	);
-
-	notify_user(
-		$entity->owner_guid,
-		$user->guid,
-		$subject,
-		$body,
-		[
-			'action' => 'create',
-			'object' => $annotation,
-			'summary' => $summary,
-			'url' => $entity->getURL(),
-		]
-	);
+$owner = $entity->getOwnerEntity();
+
+$annotation = elgg_get_annotation_from_id($annotation_id);
+
+$title_str = $entity->getDisplayName();
+if (!$title_str) {
+	$title_str = elgg_get_excerpt($entity->description, 80);
 }
 
-system_message(elgg_echo("likes:likes"));
+$site = elgg_get_site_entity();
+
+// summary for site_notifications
+$summary = elgg_echo('likes:notifications:subject', [
+		$user->name,
+		$title_str
+	],
+	$owner->language
+);
+
+// prevent long subjects in mail
+$title_str = elgg_get_excerpt($title_str, 80);
+$subject = elgg_echo('likes:notifications:subject', [
+		$user->getDisplayName(),
+		$title_str
+	],
+	$owner->language
+);
+
+$body = elgg_echo('likes:notifications:body', [
+		$owner->getDisplayName(),
+		$user->getDisplayName(),
+		$title_str,
+		$site->getDisplayName(),
+		$entity->getURL(),
+		$user->getURL()
+	],
+	$owner->language
+);
+
+notify_user(
+	$entity->owner_guid,
+	$user->guid,
+	$subject,
+	$body,
+	[
+		'action' => 'create',
+		'object' => $annotation,
+		'summary' => $summary,
+		'url' => $entity->getURL(),
+	]
+);
 
-// Forward back to the page where the user 'liked' the object
-forward(REFERER);
+return elgg_ok_response('', elgg_echo('likes:likes'));
```

Comment:
If not called by ajax, not sure this will forward. Maybe needs 3rd arg.

Reply:
it will forward to referer